pkg/audit/stats_reporter.go

```diff
@@ -0,0 +1,124 @@
+package audit
+
+import (
+	"context"
+	"time"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/metrics"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+const (
+	totalViolationsName  = "violations_total"
+	constraintsTotalName = "constraints_total"
+	auditLatency         = "audit_latency"
+	methodType           = "audit"
+)
+
+var (
+	violationsTotalM  = stats.Int64(totalViolationsName, "Total number of violations per constraint", stats.UnitNone)
+	constraintsTotalM = stats.Int64(constraintsTotalName, "Total number of enforced constraints", stats.UnitNone)
+	auditLatencyM     = stats.Float64(auditLatency, "Latency of audit operation", stats.UnitMilliseconds)
+
+	methodTypeKey     = tag.MustNewKey("method_type")
+	constraintKindKey = tag.MustNewKey("constraint_kind")
+	constraintNameKey = tag.MustNewKey("constraint_name")
+)
+
+func init() {
+	register()
+}
+
+func register() {
+	views := []*view.View{
+		{
+			Name:        totalViolationsName,
+			Measure:     violationsTotalM,
+			Aggregation: view.LastValue(),
+			TagKeys:     []tag.Key{methodTypeKey, constraintKindKey, constraintNameKey},
+		},
+		{
+			Name:        constraintsTotalName,
```

Comment:
It seems like the controller would be a better source of truth for this metric since it could keep it updated with each reconcile?

Reply:
Note that moving this to the controller also allows you to export metrics for the number of constraint templates as well as template ingestion latency.

Reply:
discussed this with @ritazh. in constraints, it is reconciled on a per constraint basis. if we were to add the client to pull this information it'll add extra processing time, which comes free in audit.

Reply:
We could have a global registry of constraints, similar to how the watch manager operates (though I don't think I'd try piggybacking off the watch manager for this).

Essentially, you have a shared map[string]bool where the keys are `{constraint_kind}/{constraint_name}` and the total number of constraints becomes len(constraints).

This is a simple (constant time) calculation that avoids the latency of audit ticks.

Reply:
I should clarify: "shared" means shared across all the individual constraint controllers.

Reply:
If that's too much for a first-pass CL we can divert this metric to a follow-on CL. That'll keep change size manageable.

(says the person with the mega-change that's clogging everything)

Reply:
I agree with Max's suggestion of maintaining the map of constraints (and constraint templates). It's near-zero cost and the gains in metric latency are huge. Also okay with deferring this to a follow up CL.

Reply:
Hey 👋 A bit late to the discussion. To ensure this map of constraints and constraint templates survive a pod restart I assume one way is to use something like `discoveryClient.ServerResourcesForGroupVersion` to get the latest state of the cluster, which is what we are already doing in audit. While I agree there is a lag in audit, does the metrics have to be realtime? (totally ok if it is, just curious what is an acceptable lag) would be nice to reuse audit since it's already doing this work anyway.  WDYT?

Reply:
Not sure what you specifically mean by "survive a pod restart".

In either the audit scraping case or the map[] case, the initial metric value after pod restart will be zero. In both cases the value will tend to converge to the number of constraints on the cluster (barring errors). The scraping case will converge after a cycle, the map[] case will converge after reconcile loops have run and constraints are added (or fail to be added).

The benefit of using the map[] interface aside from lag is that it gives closer visibility to the state of the webhook itself. Because the controllers are responsible for adding constraints to OPA, they are in the best place to sort constraints by "enforced" or "not enforced". In the end the state of the webhook (not the cluster) is what metrics are most interested in, because that tells us whether a given constraint can be enforced.

As we move further toward high-availability and begin running multiple webhooks this will allow us to do things like detect when Pods have diverging counts of constraints enforced and they will be better able to detect issues like "there is a deadlock in the controller in this updated version", that might not be visible if we rely on the `status` of a resource (which will fail because the controller isn't running). In such a case, we would see the number of constraints enforced stuck at zero and could alert, possibly stopping the rollout of a bad update (assuming a user has a deployment system capable of such a thing).

Reply:
if all the constraints have been reconciled before pod restarts, then what is the mechanism you had in mind to populate the map[]?

Reply:
I agree that the map will need to be pre-populated by an initial step when controller starts up. However, this should be trivial to implement and a very low one-time cost for the advantage of having accurate and up-to-date metrics at the source of truth. I believe you will need nearly identical code to export metrics regarding constraint templates anyways, since those are not tracked in the audit process?

Reply:
As Max reminded me, the watcher will scrape all existing resources to set the initial state. We are good. 👍

Reply:
moved totalConstraints per enforcement action to controller now. lmk if this works.